From ce9639883012b84ff1608cad8d5dbdfb2f604210 Mon Sep 17 00:00:00 2001
From: Christophe Roullier <christophe.roullier@st.com>
Date: Fri, 2 Dec 2022 10:58:52 +0100
Subject: [PATCH] net: phy: realtek: add property to disable EEE for RTL8211F

Add property "realtek,eee-disable" to have possibility to disable
EEE (Energy-Efficient Ethernet) feature in Realtek Phy.
This feature is activated by default and in some usecases (TSN, gPTP etc..)
gPTP peer delays of 18us is observed when EEE is activated.
IEEE802.1AS mention to consider a system as AScapable; the peer delay
should be around 800ns. So in this situation we are hugely out of spec
and so no gPTP compliant. Without EEE, delay is improve.

Change-Id: I587aa104f8e4d45f87b18e870cdfbf222b6c9058
Signed-off-by: Christophe Roullier <christophe.roullier@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/278943
Tested-by: Eric FOURMONT <eric.fourmont-ext@st.com>
Domain-Review: Eric FOURMONT <eric.fourmont-ext@st.com>
Reviewed-by: Eric FOURMONT <eric.fourmont-ext@st.com>
---
 drivers/net/phy/realtek.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/drivers/net/phy/realtek/realtek_main.c
+++ b/drivers/net/phy/realtek/realtek_main.c
@@ -448,6 +448,16 @@ static int rtl8211f_config_init(struct p
 			val_rxdly ? "enabled" : "disabled");
 	}
 
+	if (of_property_read_bool(dev->of_node, "realtek,eee-disable")) {
+		rtl821x_write_page(phydev, 0xa4b);
+		phy_write(phydev, 0x11, 0x1110);
+		rtl821x_write_page(phydev, 0);
+		phy_write(phydev, 0xd,7);
+		phy_write(phydev, 0xe,0x3c);
+		phy_write(phydev, 0xd,0x4007);
+		phy_write(phydev, 0xe,0x0);
+	}
+
 	if (priv->has_phycr2) {
 		ret = phy_modify_paged(phydev, 0xa43, RTL8211F_PHYCR2,
 				       RTL8211F_CLKOUT_EN, priv->phycr2);
